File: go/src/RumbleRacing/core/asset/txf/HEAD.go
package txf

import (
	"encoding/binary"
	"fmt"
)

type HEAD struct {
	Size uint32

	// Offsets in comments are offset from start of TXF File
	// this is because I reverse engineered them relative to the beginning of the TXF file.
	AllocBytes       uint16   // 0x10
	TotalTextures    uint16   // 0x12
	CLHEIterations   uint8    // 0x14
	ZTHEsCount       uint8    // 0x15
	HeadPointerCount uint8    // 0x16
	UnusedByte       uint8    // 0x17
	ZTHEFilePointers []uint32 // 4 byte absolute pointers to ZTHEs starting data within the entire TXF file. Repeats until end of HEAD section
}

func parseHEAD(buf []byte) (*HEAD, error) {

	// typical of all chunks, extract it anyway
	size := binary.LittleEndian.Uint32(buf[4:8])

	// HEAD data
	alloc := binary.LittleEndian.Uint16(buf[8:10])

	totalTextures := binary.LittleEndian.Uint16(buf[10:12])

	// missing something here
	clheIterations := buf[12]
	zthesCount := buf[13]
	headPointerCount := buf[14]
	unusedByte := buf[15]

	pointers := buf[16:]

	// fmt.Println(totalTextures, clheIterations, zthesCount, headPointerCount, unusedByte)

	if len(pointers)/4 != int(headPointerCount) {
		fmt.Println(len(pointers)/4, int(headPointerCount))
		panic("remaining head pointers byte length is not equal to the pointer count")
	}

	var ptrs []uint32

	for i := 0; i+4 <= len(pointers); i += 4 {
		nextPointer := binary.LittleEndian.Uint32(pointers[i : i+4])
		ptrs = append(ptrs, nextPointer)
	}

	// fmt.Println(hex.Dump(pointers))
	// fmt.Println(hex.Dump(buf))

	return &HEAD{
		Size:             size,
		AllocBytes:       alloc,
		TotalTextures:    totalTextures,
		CLHEIterations:   clheIterations,
		ZTHEsCount:       zthesCount,
		HeadPointerCount: headPointerCount,
		UnusedByte:       unusedByte,
		ZTHEFilePointers: ptrs,
	}, nil
}

File: src/RumbleRacing/asset/txf/HEAD.ts
import ArrayBufferSlice from "../../../ArrayBufferSlice";

export interface HEAD {
  size: number;
  allocBytes: number;
  totalTextures: number;
  clheIterations: number;
  zthesCount: number;
  headPointerCount: number;
  ztheFilePointers: number[];
}

export function parseHEAD(buf: ArrayBufferSlice): HEAD {
  const view = buf.createDataView();
  const size = view.getUint32(4, true);
  const alloc = view.getUint16(8, true);
  const totalTextures = view.getUint16(10, true);
  const clheIterations = view.getUint8(12);
  const zthesCount = view.getUint8(13);
  const headPointerCount = view.getUint8(14);

  const ptrs: number[] = [];
  for (let i = 16; i + 4 <= buf.byteLength; i += 4) {
    ptrs.push(view.getUint32(i, true));
  }

  return {
    size,
    allocBytes: alloc,
    totalTextures,
    clheIterations,
    zthesCount,
    headPointerCount,
    ztheFilePointers: ptrs,
  };
}
